jupyter_server/gateway/gateway_client.py

```diff
@@ -417,40 +420,96 @@ def load_connection_args(self, **kwargs):
         return kwargs
 
 
-async def gateway_request(endpoint, **kwargs):
+class RetryableHTTPClient:
+    """
+    Inspired by urllib.util.Retry (https://urllib3.readthedocs.io/en/stable/reference/urllib3.util.html),
+    this class is initialized with desired retry characteristics, uses a recursive method `fetch()` against an instance
+    of `AsyncHTTPClient` which tracks the current retry count across applicable request retries.
+    """
+
+    MAX_RETRIES_DEFAULT = 2
+    MAX_RETRIES_CAP = 10  # The upper limit to max_retries value.
+    max_retries: int = int(os.getenv("JUPYTER_GATEWAY_MAX_REQUEST_RETRIES", MAX_RETRIES_DEFAULT))
+    if max_retries < 0:
+        max_retries = 0
+    elif max_retries > MAX_RETRIES_CAP:
+        max_retries = MAX_RETRIES_CAP
+    retried_methods: ty.Set[str] = {"GET", "DELETE"}
+    retried_errors: ty.Set[int] = {502, 503, 504, 599}
+    retried_exceptions: ty.Set[type] = {ConnectionError}
+    backoff_factor: float = 0.1
+
+    def __init__(self):
+        self.retry_count: int = 0
+        self.client: AsyncHTTPClient = AsyncHTTPClient()
+
+    async def fetch(self, endpoint: str, **kwargs: ty.Any) -> HTTPResponse:
+        """
+        Retryable AsyncHTTPClient.fetch method.  When the request fails, this method will
+        recurse up to max_retries times if the condition deserves a retry.
+        """
+        try:
```

Comment:
I think you need to reset the retry count in the public `fetch` and have a private `_fetch` that increments it.

Reply:
Either that or make this class an implementation detail and use the `gateway_request` function as the public API (allowing this to be backward compatible).

Reply:
Ah - good catch.  The current implementation was assuming a `fetch()` call was one-to-one with the instance lifetime.